core/src/server/rpc_module.rs

```diff
@@ -738,6 +748,78 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 	}
 }
 
+/// Represent a pending subscription which waits being accepted or rejected.
+///
+/// Warning: you need to call either `PendingSubscription::accept` or `PendingSubscription::reject` otherwise
+/// the subscription will be dropped with an `InvalidParams` error.
+#[derive(Debug)]
+struct InnerPendingSubscription {
+	/// Sink.
+	sink: MethodSink,
+	/// Get notified when subscribers leave so we can exit
+	close_notify: Option<Arc<Notify>>,
+	/// MethodCallback.
+	method: &'static str,
+	/// Unique subscription.
+	uniq_sub: SubscriptionKey,
+	/// Shared Mutex of subscriptions
+	subscribers: Subscribers,
+	/// Request ID.
+	id: Id<'static>,
+}
+
+/// Represent a pending subscription which waits for being accepted or rejected.
+///
+/// This type implement drop for ease of use, such as dropped in error short circuiting via `map_err()?`.
+#[derive(Debug)]
+pub struct PendingSubscription(Option<InnerPendingSubscription>);
+
+impl PendingSubscription {
+	/// Reject the subscription call from [`ErrorObject`].
+	pub fn reject(mut self, err: impl Into<ErrorObjectOwned>) -> Result<(), Error> {
+		if let Some(inner) = self.0.take() {
+			let InnerPendingSubscription { sink, id, .. } = inner;
+
+			if sink.send_error(id, err.into()) {
+				Ok(())
+			} else {
+				Err(Error::Custom("Connection is closed".to_string()))
+			}
+		} else {
+			Err(Error::Custom("Subscription is already claimed".to_string()))
+		}
+	}
+
+	/// Attempt to accept the subscription and respond the subscription method call.
+	///
+	/// Fails if the connection was closed
+	pub fn accept(mut self) -> Result<SubscriptionSink, Error> {
+		let inner = match self.0.take() {
+			Some(inner) => inner,
+			None => return Err(Error::Custom("Subscription is already claimed".to_string())),
+		};
+
+		let InnerPendingSubscription { sink, close_notify, method, uniq_sub, subscribers, id } = inner;
+
+		if sink.send_response(id, &uniq_sub.sub_id) {
+			let active_sub = Arc::new(());
+			subscribers.lock().insert(uniq_sub.clone(), (sink.clone(), active_sub.clone()));
+			Ok(SubscriptionSink { inner: sink, close_notify, method, uniq_sub, subscribers, active_sub })
+		} else {
+			Err(Error::Custom("Connection is closed".into()))
+		}
+	}
+}
+
+impl Drop for PendingSubscription {
+	fn drop(&mut self) {
+		if let Some(inner) = self.0.take() {
```

Comment:
Ah, maybe we still need the `Option<InnerPendingSubscriptiuon>` so that the Drop impl knows whether to send anything or not; I see now!